File: core/flowcatalyst-stream-processor/src/main/java/tech/flowcatalyst/streamprocessor/postgres/EventProjectionService.java
package tech.flowcatalyst.streamprocessor.postgres;

import io.agroal.api.AgroalDataSource;
import io.quarkus.runtime.ShutdownEvent;
import io.quarkus.runtime.StartupEvent;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.event.Observes;
import jakarta.inject.Inject;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import java.sql.*;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Projects events from the events table to events_read using pure JDBC.
 *
 * <p>Polls for unprojected events (projected=false) ordered by time,
 * batch inserts to events_read, and marks them as projected.</p>
 *
 * <h2>Algorithm</h2>
 * <ol>
 *   <li>Poll: SELECT from events WHERE projected=false ORDER BY time LIMIT batchSize</li>
 *   <li>If results > 0: batch UPSERT to events_read, UPDATE events SET projected=true</li>
 *   <li>Sleep: 0ms if batchSize results, 100ms if partial, 1000ms if zero</li>
 * </ol>
 */
@ApplicationScoped
public class EventProjectionService {

    private static final Logger LOG = Logger.getLogger(EventProjectionService.class.getName());

    @Inject
    AgroalDataSource dataSource;

    @ConfigProperty(name = "stream-processor.events.enabled", defaultValue = "true")
    boolean enabled;

    @ConfigProperty(name = "stream-processor.events.batch-size", defaultValue = "100")
    int batchSize;

    private volatile boolean running = false;
    private volatile Thread pollerThread;

    void onStart(@Observes StartupEvent event) {
        if (!enabled) {
            LOG.info("Event projection service disabled");
            return;
        }
        start();
    }

    void onShutdown(@Observes ShutdownEvent event) {
        stop();
    }

    public synchronized void start() {
        if (running) {
            LOG.warning("Event projection service already running");
            return;
        }

        running = true;
        pollerThread = Thread.startVirtualThread(this::pollLoop);
        LOG.info("Event projection service started (batchSize=" + batchSize + ")");
    }

    public synchronized void stop() {
        if (!running) {
            return;
        }

        LOG.info("Stopping event projection service...");
        running = false;

        if (pollerThread != null) {
            pollerThread.interrupt();
            try {
                pollerThread.join(5000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        LOG.info("Event projection service stopped");
    }

    public boolean isRunning() {
        return running;
    }

    /**
     * Main polling loop.
     */
    private void pollLoop() {
        while (running) {
            try {
                int processed = pollAndProject();

                // Sleep based on results
                if (processed == 0) {
                    Thread.sleep(1000); // No work, sleep 1 second
                } else if (processed < batchSize) {
                    Thread.sleep(100); // Partial batch, sleep 100ms
                }
                // Full batch: no sleep, immediately poll again

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            } catch (Exception e) {
                LOG.log(Level.SEVERE, "Error in event projection poll loop", e);
                try {
                    Thread.sleep(5000); // Back off on error
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
    }

    /**
     * Poll for unprojected events and project them.
     *
     * @return number of events processed
     */
    private int pollAndProject() throws SQLException {
        List<EventRow> events = pollEvents();

        if (events.isEmpty()) {
            return 0;
        }

        try (Connection conn = dataSource.getConnection()) {
            conn.setAutoCommit(false);

            try {
                // Batch UPSERT to events_read
                upsertEventsRead(conn, events);

                // Mark as projected
                markProjected(conn, events);

                conn.commit();

                LOG.fine("Projected " + events.size() + " events");
                return events.size();

            } catch (Exception e) {
                conn.rollback();
                throw e;
            }
        }
    }

    /**
     * Poll for unprojected events.
     */
    private List<EventRow> pollEvents() throws SQLException {
        String sql = """
            SELECT id, spec_version, type, source, subject, time, data,
                   correlation_id, causation_id, deduplication_id, message_group,
                   context_data, client_id
            FROM events
            WHERE projected = false
            ORDER BY time
            LIMIT ?
            """;

        List<EventRow> events = new ArrayList<>(batchSize);

        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {

            ps.setInt(1, batchSize);

            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    events.add(mapEventRow(rs));
                }
            }
        }

        return events;
    }

    /**
     * Batch UPSERT events to events_read.
     * Note: id IS the event id (1:1 projection, no separate event_id column).
     * Note: context_data is normalized to event_read_context_data table (V12 migration).
     */
    private void upsertEventsRead(Connection conn, List<EventRow> events) throws SQLException {
        // Build multi-row INSERT with ON CONFLICT DO NOTHING
        StringBuilder sql = new StringBuilder("""
            INSERT INTO events_read (
                id, spec_version, type, source, subject, time, data,
                correlation_id, causation_id, deduplication_id, message_group,
                client_id, application, subdomain, aggregate, projected_at
            ) VALUES
            """);

        // Add value placeholders for each row (15 columns + NOW())
        for (int i = 0; i < events.size(); i++) {
            if (i > 0) sql.append(",");
            sql.append("\n(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NOW())");
        }

        sql.append("\nON CONFLICT (id) DO NOTHING");

        try (PreparedStatement ps = conn.prepareStatement(sql.toString())) {
            int paramIndex = 1;

            for (EventRow event : events) {
                // Parse type into application/subdomain/aggregate
                String[] typeSegments = event.type != null ? event.type.split(":", 4) : new String[0];
                String application = typeSegments.length > 0 ? typeSegments[0] : null;
                String subdomain = typeSegments.length > 1 ? typeSegments[1] : null;
                String aggregate = typeSegments.length > 2 ? typeSegments[2] : null;

                ps.setString(paramIndex++, event.id);           // id IS the event id
                ps.setString(paramIndex++, event.specVersion);
                ps.setString(paramIndex++, event.type);
                ps.setString(paramIndex++, event.source);
                ps.setString(paramIndex++, event.subject);
                ps.setTimestamp(paramIndex++, event.time != null ? Timestamp.from(event.time) : null);
                ps.setString(paramIndex++, event.data);
                ps.setString(paramIndex++, event.correlationId);
                ps.setString(paramIndex++, event.causationId);
                ps.setString(paramIndex++, event.deduplicationId);
                ps.setString(paramIndex++, event.messageGroup);
                ps.setString(paramIndex++, event.clientId);
                ps.setString(paramIndex++, application);
                ps.setString(paramIndex++, subdomain);
                ps.setString(paramIndex++, aggregate);
            }

            ps.executeUpdate();
        }
    }

    /**
     * Mark events as projected.
     */
    private void markProjected(Connection conn, List<EventRow> events) throws SQLException {
        // Build UPDATE with IN clause
        StringBuilder sql = new StringBuilder("UPDATE events SET projected = true WHERE id IN (");

        for (int i = 0; i < events.size(); i++) {
            if (i > 0) sql.append(",");
            sql.append("?");
        }
        sql.append(")");

        try (PreparedStatement ps = conn.prepareStatement(sql.toString())) {
            int paramIndex = 1;
            for (EventRow event : events) {
                ps.setString(paramIndex++, event.id);
            }
            ps.executeUpdate();
        }
    }

    /**
     * Map a ResultSet row to EventRow.
     */
    private EventRow mapEventRow(ResultSet rs) throws SQLException {
        return new EventRow(
            rs.getString("id"),
            rs.getString("spec_version"),
            rs.getString("type"),
            rs.getString("source"),
            rs.getString("subject"),
            rs.getTimestamp("time") != null ? rs.getTimestamp("time").toInstant() : null,
            rs.getString("data"),
            rs.getString("correlation_id"),
            rs.getString("causation_id"),
            rs.getString("deduplication_id"),
            rs.getString("message_group"),
            rs.getString("context_data"),
            rs.getString("client_id")
        );
    }

    /**
     * Internal record for event data.
     */
    private record EventRow(
        String id,
        String specVersion,
        String type,
        String source,
        String subject,
        Instant time,
        String data,
        String correlationId,
        String causationId,
        String deduplicationId,
        String messageGroup,
        String contextData,
        String clientId
    ) {}
}

File: flowcatalyst-typescript/apps/flowcatalyst/src/stream-processor/event-projection-service.ts
/**
 * Event Projection Service
 *
 * Projects events from msg_event_projection_feed to msg_events_read using pure SQL.
 *
 * Uses a single writable CTE per poll cycle: one atomic statement that
 * selects the batch, inserts into msg_events_read, and marks feed entries
 * as processed. Zero application-layer data transfer - everything stays
 * in PostgreSQL.
 *
 * Algorithm:
 *   1. Single CTE: batch SELECT -> INSERT msg_events_read -> UPDATE feed
 *   2. Sleep: 0ms if full batch, 100ms if partial, 1000ms if zero results
 *
 * Type hierarchy parsing:
 *   type: "orders:fulfillment:shipment:shipped"
 *   -> application: "orders", subdomain: "fulfillment", aggregate: "shipment"
 */

import type postgres from 'postgres';
import type { Logger } from '@flowcatalyst/logging';

export interface EventProjectionConfig {
  readonly enabled: boolean;
  readonly batchSize: number;
}

export interface EventProjectionService {
  start(): void;
  stop(): void;
  isRunning(): boolean;
}

export function createEventProjectionService(
  sql: postgres.Sql,
  config: EventProjectionConfig,
  logger: Logger,
): EventProjectionService {
  let running = false;

  /**
   * Single-statement batch projection using writable CTE.
   *
   * 1. `batch` CTE: selects unprocessed feed entries (LIMIT batchSize)
   * 2. `projected` CTE: UPSERTs into msg_events_read from batch payloads
   * 3. Main UPDATE: marks batch entries as processed
   *
   * All three operations execute atomically in one round-trip.
   * Data never leaves PostgreSQL - JSONB extraction happens in-engine.
   */
  async function pollAndProject(): Promise<number> {
    const result = await sql`
			WITH batch AS (
				SELECT id, event_id, payload
				FROM msg_event_projection_feed
				WHERE processed = 0
				ORDER BY id
				LIMIT ${config.batchSize}
			),
			projected AS (
				INSERT INTO msg_events_read (
					id, spec_version, type, source, subject, time, data,
					correlation_id, causation_id, deduplication_id, message_group,
					client_id, application, subdomain, aggregate, projected_at
				)
				SELECT
					b.event_id,
					b.payload->>'specVersion',
					b.payload->>'type',
					b.payload->>'source',
					b.payload->>'subject',
					(b.payload->>'time')::timestamptz,
					b.payload->>'data',
					b.payload->>'correlationId',
					b.payload->>'causationId',
					b.payload->>'deduplicationId',
					b.payload->>'messageGroup',
					b.payload->>'clientId',
					split_part(b.payload->>'type', ':', 1),
					NULLIF(split_part(b.payload->>'type', ':', 2), ''),
					NULLIF(split_part(b.payload->>'type', ':', 3), ''),
					NOW()
				FROM batch b
				ON CONFLICT (id) DO NOTHING
			)
			UPDATE msg_event_projection_feed
			SET processed = 1, processed_at = NOW()
			WHERE id IN (SELECT id FROM batch)
		`;

    const count = result.count;
    if (count > 0) {
      logger.debug({ count }, 'Projected events');
    }
    return count;
  }

  async function pollLoop(): Promise<void> {
    while (running) {
      try {
        const processed = await pollAndProject();

        if (processed === 0) {
          await sleep(1000); // No work, sleep 1 second
        } else if (processed < config.batchSize) {
          await sleep(100); // Partial batch, sleep 100ms
        }
        // Full batch: no sleep, immediately poll again
      } catch (err) {
        if (!running) break;
        logger.error({ err }, 'Error in event projection poll loop');
        await sleep(5000); // Back off on error
      }
    }
  }

  function start(): void {
    if (running) {
      logger.warn('Event projection service already running');
      return;
    }

    running = true;
    pollLoop().catch((err) => {
      logger.error({ err }, 'Event projection poll loop exited unexpectedly');
      running = false;
    });
    logger.info({ batchSize: config.batchSize }, 'Event projection service started');
  }

  function stop(): void {
    if (!running) return;
    logger.info('Stopping event projection service...');
    running = false;
    logger.info('Event projection service stopped');
  }

  return { start, stop, isRunning: () => running };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
